src/esp/sim/ReplayBatchRenderer.h

```diff
@@ -23,29 +22,123 @@ class Renderer;
+
+  void setEnvironmentKeyframeUnwrapped(
+      unsigned envIndex,
+      Corrade::Containers::StringView serKeyframe);
+
+  void setSensorTransform(unsigned envIndex,
+                          const std::string& sensorName,
+                          const Mn::Matrix4& transform);
+
+  // You must have done Recorder::addUserTransformToKeyframe(prefix +
+  // sensorName, ...) for every sensor in
+  // ReplayRendererConfiguration::sensorSpecifications, for the specified
+  // environment's keyframe. See also setEnvironmentKeyframe.
+  void setSensorTransformsFromKeyframe(unsigned envIndex,
+                                       const std::string& prefix);
+
+  // Renders and waits for the render to finish
+  void render(Corrade::Containers::ArrayView<const Magnum::MutableImageView2D>
+                  imageViews);
+
+  // Assumes the framebuffer color & depth is cleared
+  void render(Magnum::GL::AbstractFramebuffer& framebuffer);
+
+ private:
+  /* Implementation of all public API is in the private do*() functions,
+     similarly to how e.g. Magnum plugin interfaces work. The public API does
+     all necessary checking (such as ensuring envIndex is in bounds) in order
+     to allow the implementations be only about what's actually
+     backend-specific, with no unnecessary duplicated code. */
+
+  /* Default implementation does nothing */
+  virtual void doPreloadFile(Corrade::Containers::StringView filename);
+
+  virtual unsigned doEnvironmentCount() const = 0;
+
+  /* Retrieves a player instance for given environment. Used by
+     setSensorTransformsFromKeyframe(). The envIndex is guaranteed to be in
+     bounds. */
+  virtual esp::gfx::replay::Player& doPlayerFor(unsigned envIndex) = 0;
+
+  /* envIndex is guaranteed to be in bounds */
+  virtual Magnum::Vector2i doSensorSize(unsigned envIndex) = 0;
+
+  /* envIndex is guaranteed to be in bounds */
+  virtual void doSetSensorTransform(unsigned envIndex,
+                                    const std::string& sensorName,
+                                    const Mn::Matrix4& transform) = 0;
+
+  /* envIndex is guaranteed to be in bounds */
+  virtual void doSetSensorTransformsFromKeyframe(unsigned envIndex,
+                                                 const std::string& prefix) = 0;
+
+  /* imageViews.size() is guaranteed to be same as doEnvironmentCount() */
+  virtual void doRender(
+      Corrade::Containers::ArrayView<const Magnum::MutableImageView2D>
+          imageViews) = 0;
+
+  virtual void doRender(Magnum::GL::AbstractFramebuffer& framebuffer) = 0;
+
+  ESP_SMART_POINTERS(AbstractReplayRenderer)
 };
 
-class ReplayBatchRenderer {
+class ReplayRenderer : public AbstractReplayRenderer {
  public:
-  class EnvironmentRecord {
-   public:
+  struct EnvironmentRecord {
+    explicit EnvironmentRecord(
+        std::unique_ptr<gfx::replay::AbstractSceneGraphPlayerImplementation>
+            playerImplementation)
```

Comment:
```suggestion
        std::unique_ptr<gfx::replay::AbstractSceneGraphPlayerImplementation>&&
            playerImplementation)
```

Reply:
Does this actually matter? It's not like the type would be copyable, one has to `std::move()` it always.